### ISSUE TITLE
Duplicate class found

### ISSUE THREAD

**Issue by piet8stevens** [Carleslc/Simple-YAML#41]
Hi, thank you for your Simple-YAML module. I am trying to use this in my app but run into gradle dependency issues as follows:
- I have included this in my build.gradle file: 
implementation 'me.carleslc.Simple-YAML:Simple-Yaml:1.7.2'
- when I build the app, I receive this: 
Duplicate class org.simpleyaml.configuration.Configuration found in modules jetified-Simple-Configuration-1.7.2 (me.carleslc.Simple-YAML:Simple-Configuration:1.7.2) and jetified-Simple-Yaml-1.7.2 (me.carleslc.Simple-YAML:Simple-Yaml:1.7.2)

I am not familiar enough with gradle to know how to exclude such a duplicate class. 

Any help would be much appreciated. Thank you.

**Comment by portlek** [comment]
could you show your gradle file?

**Comment by piet8stevens** [comment]
Thank you for the fast reply. See attached.file:///mnt/SS/pstevens/workspace.android-studio/C-a/app/build.gradle

**Comment by portlek** [comment]
```gradle
apply plugin: 'com.android.application'

def getGitHash = { ->
    def stdout = new ByteArrayOutputStream()
    exec {
        commandLine 'git', 'rev-parse', '--short=8', 'HEAD'
        standardOutput = stdout
    }
    return stdout.toString().trim()
}

android {
    compileSdkVersion 28
    buildToolsVersion '29.0.3'
    defaultConfig {
        applicationId "com.pas.ipd_1"
        minSdkVersion 28
        targetSdkVersion 28
        versionCode 232
        versionName "5.6.0"
        buildConfigField "String", "GitHash", "\"${getGitHash()}\""

        testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
    }
    repositories {
        google()
        maven { url 'https://jitpack.io' }
    }

    android.buildFeatures.viewBinding = true

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
        debug {
            minifyEnabled false
//            testCoverageEnabled = true
        }
    }
    productFlavors {
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
    testOptions {
        unitTests {
            includeAndroidResources = true
        }
    }
    packagingOptions {
        exclude 'README.txt'
        exclude 'META-INF/LICENSE.md'
        exclude 'META-INF/LICENSE-notice.md'

    }
}

dependencies {
    implementation 'androidx.constraintlayout:constraintlayout:2.0.4'
    implementation 'androidx.test.espresso:espresso-idling-resource:3.3.0'
    implementation project(':FastBleLib')
    implementation project(path: ':MultiSelectSpinner')
    implementation 'org.junit.vintage:junit-vintage-engine:5.7.0'
    implementation 'me.carleslc.Simple-YAML:Simple-Yaml:1.7.2'
    implementation 'junit:junit:4.13.1'

    testImplementation 'junit:junit:4.13.1'
    testImplementation 'org.robolectric:robolectric:4.4'
    testImplementation 'org.junit.jupiter:junit-jupiter:5.7.0'

    androidTestImplementation 'junit:junit:4.13.1'
    androidTestImplementation 'androidx.test:core:1.3.0'

    // AndroidJUnitRunner and JUnit Rules
    androidTestImplementation 'androidx.test:runner:1.3.0'
    androidTestImplementation 'androidx.test:rules:1.3.0'

    androidTestImplementation 'com.jayway.android.robotium:robotium-solo:5.6.3'
    androidTestImplementation 'org.awaitility:awaitility:4.0.3'
//    androidTestImplementation 'org.hamcrest:hamcrest:2.1' // in order to prevent a dependency version conflict (pray)
//    androidTestImplementation 'org.hamcrest:hamcrest-core:2.1'
//    androidTestImplementation 'org.hamcrest:hamcrest-core:1.3'
//    configurations.all {
//        resolutionStrategy {
//            force 'org.hamcrest:hamcrest-core:2.1', 'org.hamcrest:hamcrest-library:2.1'
//        }
//    }
    // Assertions
    androidTestImplementation 'androidx.test.ext:junit:1.1.2'
    androidTestImplementation 'androidx.test.ext:truth:1.3.0'
    androidTestImplementation 'com.google.truth:truth:1.1'

    // Espresso dependencies
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
    androidTestImplementation 'androidx.test.espresso:espresso-contrib:3.3.0'
    androidTestImplementation 'androidx.test.espresso:espresso-intents:3.3.0'
    androidTestImplementation 'androidx.test.espresso:espresso-accessibility:3.3.0'
    androidTestImplementation 'androidx.test.espresso:espresso-web:3.3.0'
    androidTestImplementation 'androidx.test.espresso.idling:idling-concurrent:3.3.0'

    // The following Espresso dependency can be either "implementation"
    // or "androidTestImplementation", depending on whether you want the
    // dependency to appear on your APK's compile classpath or the test APK
    // classpath.
    androidTestImplementation 'androidx.test.espresso:espresso-idling-resource:3.3.0'
// https://mvnrepository.com/artifact/org.robolectric/android-all
    testImplementation group: 'org.robolectric', name: 'android-all', version: '10-robolectric-5803371
}
```

**Comment by portlek** [comment]
I guess you didn't add the jitpack as repository

**Comment by portlek** [comment]
maybe you can try to move repositories out.
![image](https://user-images.githubusercontent.com/12835420/101276807-9f40ba80-37c0-11eb-9c14-6236b1ab1f0f.png)
repositories {
  maven { url 'https://jitpack.io' }
}

**Comment by piet8stevens** [comment]
I have modified my build.gradle per your recommendation: 
[build.gradle.txt](https://github.com/Carleslc/Simple-YAML/files/5648553/build.gradle.txt)

But I still have the same problem when I build.

**Comment by piet8stevens** [comment]
I am not sure if this helps: I tried to follow the gradle recommendations on how to find conflicting dependencies. I use android studio and looked up one of the class names that came from the detailed error messages when I try to run my application (from Android studio). I did that for class TagToken, and it gives this:
![Screenshot-20201206105201-1443x261](https://user-images.githubusercontent.com/1836276/101276983-2dad4000-37b1-11eb-8c27-c33d3beb0475.png)

**Comment by portlek** [comment]
so you can exclude snakeyaml from simple-yaml dependency. probably it solves your problem.
https://docs.gradle.org/current/userguide/dependency_downgrade_and_exclude.html#sec:excluding-transitive-deps

**Comment by piet8stevens** [comment]
Yes, I had read that before contacting you and tried to put that in my build.gradle, but it results in error messages. Sorry to bother you with this.

I tried this:
    implementation 'me.carleslc.Simple-YAML:Simple-Yaml:1.7.2' {
        exclude group: 'org.yaml', module: 'snakeyaml'
    }

But it is not correct and results in this error message:
Could not find method me.carleslc.Simple-YAML:Simple-Yaml:1.7.2() for arguments [build_f1pdfqravsygqjtzxjrwlf6e$_run_closure3$_closure16@2330b009] on object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler

**Comment by portlek** [comment]
did you try that?
```gradle
implementation('me.carleslc.Simple-YAML:Simple-Yaml:1.7.2') {
  exclude group: 'org.yaml', module: 'snakeyaml'
}
```

**Comment by piet8stevens** [comment]
I have now put in the parentheses but still get the same error message:
Duplicate class org.simpleyaml.configuration.Configuration found in modules jetified-Simple-Configuration-1.7.2 (me.carleslc.Simple-YAML:Simple-Configuration:1.7.2) and jetified-Simple-Yaml-1.7.2 (me.carleslc.Simple-YAML:Simple-Yaml:1.7.2)

**Comment by portlek** [comment]
yea because simple-configuration has not snakeyaml. but I couldn't understand why the excluding is not working, interesting...

**Comment by piet8stevens** [comment]
It seems to me that the exclude works:
without exclude: duplication between Simple-Yaml and snakeyaml
     Duplicate class org.yaml.snakeyaml.util.ArrayUtils found in modules jetified-Simple-Yaml-1.7.2 (me.carleslc.Simple-YAML:Simple-Yaml:1.7.2) and jetified-snakeyaml-1.26 (org.yaml:snakeyaml:1.26)
     
with exclude: duplication between Simple-Configuration and Simple-Yaml
     Duplicate class org.simpleyaml.configuration.Configuration found in modules jetified-Simple-Configuration-1.7.2 (me.carleslc.Simple-YAML:Simple-Configuration:1.7.2) and jetified-Simple-Yaml-1.7.2 (me.carleslc.Simple-YAML:Simple-Yaml:1.7.2)     
    
So, there is an effect of the exclude, but the result is a different error.

**Comment by piet8stevens** [comment]
I have been able to solve this (at least I can run my app) by having this exclude:
    implementation('me.carleslc.Simple-YAML:Simple-Yaml:1.7.2') {
        exclude group: 'org.yaml', module: 'snakeyaml'
        exclude group: 'me.carleslc.Simple-YAML', module: 'Simple-Configuration'
    }

I very much appreciate your (very fast) support in this matter. I believe the exclude of Simple-Configuration from Simple-YAML is at best a temporary work-around. 

I will now continue to debug my own stuff. If you have time to figure out why Simple-Configuration module must be excluded, I would appreciate a message so I can understand better what is happening. I must admit I am not confident that above solution will work in the end, but I still have my own bugs to fix before I can confirm whether it works.

**Comment by portlek** [comment]
I created a new android project, it worked well. I've not seen any error.

**Comment by piet8stevens** [comment]
That is an excellent idea. I did the same. I created a new project in Android studio. I still have the  problem with duplication.  This is the gradle file: 
[build.gradle.txt](https://github.com/Carleslc/Simple-YAML/files/5648767/build.gradle.txt)

This is the error I get: Duplicate class org.simpleyaml.configuration.Configuration found in modules jetified-Simple-Configuration-1.7.2 (me.carleslc.Simple-YAML:Simple-Configuration:1.7.2) and jetified-Simple-Yaml-1.7.2 (me.carleslc.Simple-YAML:Simple-Yaml:1.7.2)
